odoo/tools/populate.py

```diff
@@ -0,0 +1,357 @@
+            comodel_max_id = populated[comodel]
+            # we use MOD() instead of %, because % cannot be correctly escaped, it's a limitation of the SQL wrapper
+            return _field.name, SQL("%(table_alias)s.%(field_name)s + %(comodel_max_id)s * (MOD(%(series_alias)s - 1, %(factor)s) + 1)",
+                                    table_alias=SQL.identifier(table_alias),
+                                    field_name=SQL.identifier(_field.name),
+                                    comodel_max_id=comodel_max_id,
+                                    series_alias=SQL.identifier(series_alias),
+                                    factor=factors[comodel])
+        return copy(_field)
+
+    if field.name == 'id':
+        return copy_id()
+    match field.type:
+        case 'one2many':
+            # there is nothing to copy, as it's value is implicitly read from the inverse Many2one
+            return copy_noop()
+        case 'many2many':
+            # there is nothing to do, the copying of the m2m will be handled when copying the relation table
+            return copy_noop()
+        case 'many2one':
+            return copy_many2one(field)
+        case 'many2one_reference':
+            # TODO: in the case of a reference field, there is no comodel,
+            #  but it's specified as the value of the field specified by model_field.
+            #  Not really sure how to handle this, as it involves reading the content pointed by model_field
+            #  to check on-the-fly if it's populated or not python-side, so for now we raw-copy it.
+            #  If we need to read on-the-fly, the populated structure needs to be in DB (via a new Model?)
+            return copy(field)
+        case 'binary':
+            # copy only binary field that are inlined in the table
+            return copy(field) if not field.attachment else copy_noop()
+        case _:
+            return copy(field)
+
+
+def populate_model(env, model, populated, factors, separator_code):
+
+    def update_sequence(_model):
+        env.execute_query(SQL("SELECT SETVAL(%(sequence)s, %(last_id)s, TRUE)",
+                              sequence=f"{_model._table}_id_seq", last_id=fetch_last_id(env, _model)))
+
+    def has_column(_field):
+        return _field.store and _field.column_type
+
+    assert model not in populated, f"We do not populate a model ({model}) that has already been populated."
+    _logger.info('Populating model %s %s times...', model._name, factors[model])
+    dest_fields = []
+    src_fields = []
+    update_fields = []
+    table_alias = 't'
+    series_alias = 's'
+    # process all stored fields (that has a respective column), if the model has an 'id', it's processed first
+    for field in (f for f in sorted(model._fields.values(), key=lambda x: x.name != 'id') if has_column(f)):
+        if field_need_variation(env, model, field) and field.type in ('char', 'text'):
+            update_fields += [field]
```

Comment:
Why do you need to update only char fields?
I thought the tool was supposed to only add new rows. Some more general documentation should be added for this entire file.

Reply:
> Why do you need to update only char fields?
I thought the tool was supposed to only add new rows.

It's a workaround to allow the command to be re-entrant.
Let's take an example with res.company.name, which has an unique constraint on it.
Initially, we were just adding the series to the names of the new records. So on the first run you had:
- `My Company`  <-- original record.
- `My Company 1` <-- + series = 1
- `My Company 2` <-- + series = 2
...
Now let's say you redo the command:
- `My Company`   <-- original record
- `My Company 1` <-- record from first run
- `My Company 2` <-- record from first run
- `My Company 1` <-- original record + series = 1 ⚠️ conflict with `My Company 1` from the first run.

So the workaround we landed on was to 'mark' each run of the populate, by appending any symbol that serves as a marker.
For ex: if sep='_', the above example becomes:
- `My Company_`   <-- original record
- `My Company_1` <-- record from first run
- `My Company_2` <-- record from first run
- `My Company__1` <-- original record + sep + series = 1, no conflict with `My Company_1` (note the double `_`).
Any symbol serves the purpose, that's why it's left as choice at CLI arg.

I hope with this it's more clear.

> Some more general documentation should be added for this entire file.

What kind of general documentation are you expecting? A short overview of the algorithm? (I've been looking at this code a quite some time, so I'm biased and not sure what another dev might find unintuitive/unclear, so any input is welcome.)

Reply:
For the documentation, I was looking for a general description of the algorithm (can be in the header of the file).

Indeed, this way you can execute populate again.